command/agent/cache/handler.go

```diff
@@ -37,8 +37,11 @@ func Handler(ctx context.Context, logger hclog.Logger, proxier Proxier, inmemSin
 		// Parse and reset body.
 		reqBody, err := ioutil.ReadAll(r.Body)
 		if err != nil {
-			logger.Error("failed to read request body")
-			logical.RespondError(w, http.StatusInternalServerError, errors.New("failed to read request body"))
+			errRet := errors.New("failed to read request body")
```

Comment:
so, logical.RespondError calls logical. AdjustErrorStatusCode in which the status code might change. I moved AdjustErrorStatusCode out of logical.RespondError so that I would not need to pass in a core reference to the logical package as it could cause circular dependency.

Reply:
Sorry, I don't follow.  How is AdjustErrorStatusCode dependent on Core?

Reply:
Sorry for my bad explanation. So, AdjustErrorStatusCode does not depend on Core. However, to set the headers based on a status code, I need to first call AdjustErrorStatusCode and then call the function that sets the headers. The main reason I move AdjustErrorStatusCode out of logical.RespondError is such that I would not need to pass in that function anything extra specially since it is in the sdk/logical package. 
A little bit more context, so, the way that setting custom headers was implemented before was that I passed in a reference to Core and through that I called the SetCustomResponseHeaders. This would mean that I would need to pass in Core to the logical.RespondError which caused circular dependency. However, now I send the core.SetCustomResponseHeaders function itself, and I can define a function type to avoid the circular dependency issue. But, wondering would the change (moving out the AdjustErrorStatusCode out of logical.RespondError) still a bad Idea? 
In any case, I think the decision on this depends on the other comments about wrapping ResponseWriter, or including the headers in the http request context.